lib/gossip/damper.js

```diff
@@ -0,0 +1,542 @@
+        return scores.every(function each(score) {
+            return score.dampScore >= suppressLimit;
+        });
+    }
+}
+
+// This function definition is pulled up and attached to the constructor
+// so that its output can be easily unit tested.
+Damper.getMembersToDamp = function getMembersToDamp(allResponses, rVal, config) {
+    var dampScoresByFlapper = groupDampScoresByFlapper(allResponses);
+    var suppressLimit = config.get('dampScoringSuppressLimit');
+    return findFlappersToDamp(dampScoresByFlapper, rVal, suppressLimit);
+};
+
+Damper.prototype.addDampedMember = function addDampedMember(memberAddr) {
+    this.dampedMembers[memberAddr] = {
+        timestamp: this.Date.now()
+    };
+};
+
+Damper.prototype.addFlapper = function addFlapper(flapper) {
+    if (this.dampedMembers[flapper.address]) {
+        this.logger.debug('ringpop damper already damped member', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    if (this.flappers[flapper.address]) {
+        this.logger.debug('ringpop damper already added flapper', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    // TODO Limit number of flappers?
+    // TODO Expire flappers?
+
+    this.flappers[flapper.address] = flapper;
+    this.logger.debug('ringpop damper added flapper', {
+        local: this.ringpop.whoami(),
+        flapper: flapper.address
+    });
+    this.ringpop.stat('increment', 'damper.flapper.added');
+    this.ringpop.stat('gauge', 'damper.flappers',
+        Object.keys(this.flappers.length));
+
+    if (this._getFlapperCount() === 1) {
+        this._startDampTimer();
+    }
+
+    return true;
+};
+
+Damper.prototype.dampMember = function dampMember(addr)  {
+    var self = this;
+    var member = this.ringpop.membership.findMemberByAddress(addr);
+
+    if (!member) {
+        this.logger.warn('ringpop damper cannot damp member; member does not exist', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    if (!this.hasFlapper(member)) {
+        this.logger.warn('ringpop damper cannot damp member; member is not flappy', {
+            local: this.ringpop.whoami(),
+            member: addr
+        });
+        return false;
+    }
+
+    this.ringpop.membership.makeDamped(member.address);
+
+    // Don't start anymore damping subprotocols for a member
+    // that is already damped.
+    this.removeFlapper(member.address);
+    this.addDampedMember(member.address);
+
+    if (!this.expirationTimer) {
+        scheduleExpiration();
+        this.logger.info('ringpop started damped member expiration timer', {
+            local: this.ringpop.whoami()
+        });
+    }
+
+    return true;
+
+    function scheduleExpiration() {
+        self.expirationTimer =
+            self.timers.setTimeout(function onTimeout() {
+                self.expireDampedMembers();
+                scheduleExpiration();
```

Comment:
I think I understand what you're trying to get at, but since `expireDampedMembers()` is blocking/synchronous, the timer will either be cancelled/cleared by `_stopExpirationTimer`. Or am I misunderstanding?